### PR TITLE
[8.0] mrp_hook + mrp_operations_extension

### PR DESCRIPTION
**Pull request by pedrobaeza**
This PR includes 2 modules because the first is only a technical one for providing hooks for allowing to work properly the other one (and others that will come later).

This is the README for the main one:
# 
# MRP operations extension module

This module adds:
- New table to store operations to avoid typing them again.
- Adds a relation from workcenter lines to BoM lists.
- Adds a relation from workcenter lines to Manufacturing Orders in
  Scheduled/Consumed/Finished products.
- Adds a relation between routing workcenter lines and workcenter extra info.
- Allows to set specific times per routing in addition to workcenter times.
- Controls the availability of material in operations
- Controls operation beginning in previous operations are not finished.
# Configuration

Go to _Settings > Manufacturing_ and activate "Manage routings and work orders"
for handling the features of this module.

### CODE REVIEW

**Comment by oihane** [comment]
@pedrobaeza demo data is causing travis fail:

```
ParseError: "hr.employee" while parsing /home/travis/build/OCA/manufacture/mrp_operations_extension/demo/hr_employee_demo.xml:5, near
<record id="hr.employee_al" model="hr.employee">
    <field name="user_id" ref="mrp_operations_extension.user_opeext_1"/>
</record>
```

**Comment by oihane** [comment]
**mrp_operation_extension** requires dependency with **hr**

**Comment by rafaelbn** [comment]
Could anyone add label need review please?

**Comment by nhomar** [comment]
Hello thanks.

Comment 1.- Please check which code is not tested in order to a little tests to achieve  a better coverage.

File: mrp_hook/models/mrp_bom.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+from openerp import models, api, tools, _
+from openerp.exceptions import Warning as UserError
+from openerp.addons.product import _common
+
+
+class MrpBom(models.Model):
+    _inherit = 'mrp.bom'
+
+    @api.model
+    def _factor(self, factor, product_efficiency, product_rounding):
```
Comment (nhomar) [inline mrp_hook/models/mrp_bom.py:14]: Can you document this a little better, I understand that you can force a factor, but "why"?

factor / NUMBER NEAR TO 0 == INFINITE , why the denominator is 1.0 should not be forced the factor = 0?

Comment (pedrobaeza) [inline mrp_hook/models/mrp_bom.py:14]: This is a verbatim copy of the core one

File: mrp_operations_extension/__init__.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        routing_wcs = env['mrp.routing.workcenter'].search(
+            [('op_wc_lines', '=', False)])
+        for routing_wc in routing_wcs:
+            routing_wc.op_wc_lines = [
+                (0, 0, {'workcenter': routing_wc.workcenter_id,
+                        'default': True,
+                        'custom_data': False})]
+
+
+def post_init_hook(cr, pool):
```
Comment (nhomar) [inline mrp_operations_extension/__init__.py]: Can you documment this method to toe README in order to inform users which are checking the documentation that the data structure and the real data will change once this module is installed?.

**Comment by nhomar** [comment]
It seems a cool set of modules conceptually +1.

I should recommend several things in order to ensure better review process:

1.- A little video with the before - after this 2 modules.
2.- More tests, I think there are too little test cases (but this is an impression), I think this 2 modules are like the base of everything then we need to chck the traceback even the init hook should be tested, becaus in an error this module will change a lot data in an actual MRP environment.
3.- The init hook should have a backup strategy in order to ensure data changed is backuped (this should be an extra not mandatory for my +1)

Thanks a lot for this kind of work.

File: mrp_hook/models/mrp_bom.py (hunk 1 above)
Comment (nhomar) [inline mrp_hook/models/mrp_bom.py:14]: Cool, can you put that on the docstring please?:

```
verbatin copy from here _"link to code"
```

Comment (pedrobaeza) [inline mrp_hook/models/mrp_bom.py:14]: Well, the entire module is a verbatim copy with the corresponding hooks. That's why there isn't tests (not in this module), and the documentation is in the README, but I can be more explicit..

**Comment by moylop260** [comment]
Could you check next [result build](https://travis-ci.org/OCA/manufacture/jobs/93542578#L266-L267)?

``` bash
************* Module mrp_hook.__openerp__
mrp_hook/__openerp__.py:5: [C8101(manifest-required-author), ] Missing author required "Odoo Community Association (OCA)" in manifest file
mrp_hook/__openerp__.py:5: [C8102(manifest-required-key), ] Missing required key "license" in manifest file
mrp_hook/__openerp__.py:5: [C8102(manifest-required-key), ] Missing required key "images" in manifest file
************* Module mrp_operations_extension
mrp_operations_extension/__init__.py:1: [W7902(duplicate-xml-record-id), ] Duplicate xml record id mrp_operations_extension.req_link_mrp_workcenter.noupdate=0
************* Module mrp_operations_extension.__openerp__
mrp_operations_extension/__openerp__.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`
mrp_operations_extension/__openerp__.py:19: [C8102(manifest-required-key), ] Missing required key "license" in manifest file
mrp_operations_extension/__openerp__.py:19: [C8102(manifest-required-key), ] Missing required key "images" in manifest file
```

**Comment by JMODaERP** [comment]
Hello Pedro:

Step 1: Add a relation from workcenter lines to BoM lists.

**I have done step 1 and have enabled the BOM**.

Step 2:  Add a relation from workcenter lines to Manufacturing Orders in Scheduled/Consumed/Finished products.

**What do you mean: Adds a relation from workcenter lines to Manufacturing Orders in Scheduled / Consumed / Finished products ?**. When I click "Compute Data", the system automatically enters the workplace and the need to produce material.

File: mrp_hook/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    "version": "8.0.1.0.0",
+    "category": "Hidden",
+    "author": "OdooMRP team, "
+              "AvanzOSC, "
+              "Serv. Tecnol. Avanzados - Pedro M. Baeza",
+    "website": "http://www.odoomrp.com",
+    "contributors": [
+        "Pedro M. Baeza <pedro.baeza@serviciosbaeza.com>",
+    ],
+    "depends": [
```
Comment (elicoidal) [inline mrp_hook/__openerp__.py:19]: Please remove empty keys

File: mrp_hook/models/mrp_bom.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Serv. Tecnol. Avanzados - Pedro M. Baeza
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp import models, api, tools, _
```
Comment (elicoidal) [inline mrp_hook/models/mrp_bom.py]: Please sort in alphabetical order following: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#imports

File: mrp_operations_extension/README.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+   :alt: License: AGPL-3
+
+===============================
+MRP operations extension module
+===============================
+
+This module adds:
+
+* New table to store operations to avoid typing them again.
+* Adds a relation from workcenter lines to BoM lists.
```
Comment (elicoidal) [inline mrp_operations_extension/README.rst]: "A relation between workcenter lines and BoM lists."

File: mrp_operations_extension/README.rst [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+* Adds a relation from workcenter lines to Manufacturing Orders in
```
Comment (elicoidal) [inline mrp_operations_extension/README.rst]: "A relation between workcenter lines and Manufacturing Orders in
  Scheduled/Consumed/Finished products."

**Comment by JMODaERP** [comment]
Well, talking about relations in Odoo, for each product there are one and only one bill of materials.

2015-11-29 19_13_36-sin titulo - paint

A Bill of Row Materials (BoM) can contain one o more Row Material, but only one Routing (production line), we could say that BoM is one table. Whose line of production is only and its primary key is the product. The BoM contains one foreing key for each row material an only one for routing (contains one foreing key for each row material and only one for routing (routing, productive line and productive process are the same concept).

In Odoo, Work Order is a relation between Operation and end product, both takes its name. Wok order is performed in a Work Center, and the capacity of the Work Center is predefined in itself, unless, otherwise directed by a work order.

Therefore a table is necessary or at least a view to store and then display, the actual capacity of the Centre for a work order and the hours provided for her in particular.

For example, if the machinery of a work center consists of a saw, does not require the same effort cutting timber that polystyrene foams, and therefore has less real capacity for the first thing for the second and that means more running time for the same operation, cut.

Do now, again I ask, what I have to prove?. What test requires the PR

**Comment by anajuaristi** [comment]
Hi Jose Manuel:

The following sentence is not true... for each product there are one and
only one bill of materials. Relation is one product, several bill of
materials, where each bill of materials contains a routing.
Routing contains routing steps. In standar Odoo one routing step contains a
single machine but in OdooMRP we added the option to have several machines
on each routing step containing each one their own capacity related to a
routing line.

Odoo standar structure has been strongly extended by OdooMRP so IMHO your
relational graphic is not exactly the one we have.

Wishing this helps:

Ana

2015-11-29 19:22 GMT+01:00 José Manuel Otero Dapena <
notifications@github.com>:

> Well, talking about relations in Odoo, for each product there are one and
> only one bill of materials.
> 
> [image: 2015-11-29 19_13_36-sin titulo - paint]
> https://cloud.githubusercontent.com/assets/15608043/11459079/72676c40-96ce-11e5-81b9-22437860722e.jpg
> 
> A Bill of Row Materials (BoM) can contain one o more Row Material, but
> only one Routing (production line), we could say that BoM is one table.
> Whose line of production is only and its primary key is the product. The
> BoM contains one foreing key for each row material an only one for routing
> (contains one foreing key for each row material and only one for routing
> (routing, productive line and productive process are the same concept).
> 
> In Odoo, Work Order is a relation between Operation and end product, both
> takes its name. Wok order is performed in a Work Center, and the capacity
> of the Work Center is predefined in itself, unless, otherwise directed by a
> work order.
> 
> Therefore a table is necessary or at least a view to store and then
> display, the actual capacity of the Centre for a work order and the hours
> provided for her in particular.
> 
> For example, if the machinery of a work center consists of a saw, does not
> require the same effort cutting timber that polystyrene foams, and
> therefore has less real capacity for the first thing for the second and
> that means more running time for the same operation, cut.
> 
> Do now, again I ask, what I have to prove?. What test requires the PR
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/manufacture/pull/57#issuecomment-160442214.

## 

CEO Avanzosc, S.L http://www.avanzosc.es : Office phone / Tfono oficina:
(+34) 943 02 69 02
Ana Juaristi Olalde http://www.anajuaristi.com/: Personal phone: 677 93
42 59. User/usuario skype: Avanzosc
www.openerpsite.com

_El contenido de esta comunicación y de toda su documentación anexa es
confidencial y se dirige exclusivamente a su destinatario. El uso no
autorizado de esta información está prohibido por la legislación vigente.
Si usted no es el destinatario le rogamos nos lo indique, no comunique su
contenido a terceros y proceda a su destrucción. Disculpe las molestias que
le haya ocasionado la recepción indebida de este e-mail. Sus datos figuran
en un fichero cuyo titular es Avanzosc, S.L., a quien usted puede dirigirse
para ejercer sus derechos de acceso, rectificación, cancelación y oposición
en Klara Donea 13, 20720, Azkoitia (Gipuzkoa), Tef. 943 02 69 02 -
*_administracion@avanzosc.
soporte@avanzosc.eses*

_Komunikazio honen edukia eta dokumentazio erantsia konfidentziala da eta
hartzaileak bakarrik jaso beharko luke. Indarrean dagoen legeriak debekatu
egiten du bertan eskainitako informazioa baimenik gabe erabiltzea.
Komunikazioa zuri iritsi bazaizu, baina zu ez bazara hartzailea, mesedez,
guri jakinarazi, eta jasotako informazioa ez inori jakinarazi eta suntsitu.
Barkatu okerreko email hau jasotzeak eragindako eragozpenak. Zure datuak
Avanzosc, S.L. enpresaren fitxategietan sartuta daude. Zure datuak atzitzea
eska dezakezu, bai eta, datuak zuzentzea, ezereztea eta tratamenduari aurka
egitea ere. Horretarako, enpresara jo dezakezu, helbide honetan: *_Klara
Donea 13*\* 20720, Azkoitia (Gipuzkoa), telefonoa: 943 02 69 02 - *
_administracion@avanzosc. soporte@avanzosc.eses_ _This message and all
documents attached to it are confidential and intended only for the person
or entity to which it is addressed. Any use of this information by
unauthorised persons is prohibited under current legislation. If you
received this message by error, please advise us, destroy it and refrain
from communicating its contents to third parties. We apologise for any
inconvenience receiving this email improperly may cause to you. Your
personal data are included in a file owned by Avanzosc, S.L. If you want to
exercise your rights of access, correction, erasure and objection you can
contact the Controller at *_Klara Donea 13*\* 20720, Azkoitia (Gipuzkoa), T:
943 02 69 02 – administracion@avanzosc. soporte@avanzosc.eses*

**Comment by JMODaERP** [comment]
You're right in the foregoing. 

As we can see is configurable: “Allow several bill of materials per products using properties”. Until now I have not used this feature.

As for the rest, I know that in a customized version several resources (humans and machines) can be used, to the end and both are within productive resources, and is logical that each resource has a different potential capacity. I've tested Odoo Core, Odoo OCA and Customer Odoo and to be several possibilities I've painted the simplest one. 

Your notes have been very useful. I shall remake  the diagram of entities and modify the relations. 

A cordial greeting.

**Comment by rafaelbn** [comment]
@anajuaristi @pedrobaeza about @nhomar https://github.com/OCA/manufacture/pull/57#issuecomment-160173847
- Ana do you have any short video? or something similar?
- Pedro, I think that if you review comments, Eva, Ana and me could make las functional review and go ahead with this.

ping @tafaRU

**Comment by rafaelbn** [comment]
In mrp_operation_extension's README there is a mistake in configuration:

> Go to Settings > Manufacturing and activate "Manage routings and work orders" for handling the features of this module.
- Activating this option doesn't install mrp_operation_extension

I would recommend for testing: 
1. Start a base database


2. Go To Setting - User and activate technical features in admin user
3. Go To Setting - Modules and install **only**  mrp_operations_extension 
4. Go to Setting - Manufacturing - Planning and enable options
    - Manage routing and work order
    - Allow detailed planning of work order
    - Manage operators in work centers

**Comment by ybofr** [comment]
Hi,
Is the pull request available for testing on OCA? if yes, what should I choose, as I presume, I should not choose OdooMRP anymore.

Thanks for the answer
Eva P.

**Comment by oihane** [comment]
Hi @ybofr,
can you see the checks at the end of the PR?

image

There is one that says **runbot** if you click on details, you will be redirected to the corresponding OCA runbot

**Comment by anajuaristi** [comment]
I would make a sugestion... 
Manage routing and workcenter is an Odoo core check that installs mrp_operations
I think if someone installs directly mrp_operations_extension, that check should be automatically enabled since it has got a dependancy with mrp_operations.

It has nosense installing mrp_operations_extension and don't using routings. 
My2cents

**Comment by pedrobaeza** [comment]
@rafaelbn, don't lose the focus: configuration instructions are once you have installed the module, so _mrp_operations_ and _mrp_operations_extension_ are already installed in this point. You only need to set the check for viewing all the stuff.

@anajuaristi, the check doesn't install the module. It simply un-hide some fields and menus.

**Comment by anajuaristi** [comment]
Yes, maybe I did not explain very well

I wanted to say that if someone installs operations_extension, that check should be automatically enabled. Otherwise it must to be made manually, since a lot of fields are invisible if it's not enabled. It will improve usability for user. 
Just a sugestion.
.

**Comment by pedrobaeza** [comment]
Well, that's up to Odoo that doesn't enable it by default installing _mrp_operations_.

**Comment by JavierIniesta** [comment]
Tested in runbot :+1:

**Comment by JMODaERP** [comment]
Tested in runbot, :+1:

I am making a little visual procces for explain the proccess to prove.

**Comment by rafaelbn** [comment]
@pedrobaeza could you review comment from @elicoidal please?

**Comment by rafaelbn** [comment]
Hi @ybofr you could testit here: http://runbot.odoo-community.org/runbot/build/3136482

**Comment by anajuaristi** [comment]
Tested in
http://runbot.odoo-community.org/runbot/build/3136482

Steps to reproduce:

Create a new product "production"
Creat a new bom for it
Include existing materials on bom
Activate the bom
Go to manufacturing orders from product "production"
Push compute data
Try to confirm
ValidateError

Field(s) `product_uom` failed against a constraint: You try to move a product using a UoM that is not compatible with the UoM of the product moved. Please use an UoM in the same UoM category.

Activate "use different measure units" just to see witch of them is different one.
All materials are set on UOM "unit(s)"

I tried also to set a cost product "service" for machines but I can not find the place where the unit is wrong so I can not confirm the order.

Sorry but having this behavior I have got to say :-1: until we find the wrong product and system gives more clues about how to solve the error.

**Comment by oihane** [comment]
@anajuaristi did you test in **-all** database or in **-base** database?

**Comment by anajuaristi** [comment]
In the one i provide the link. The same used by @rafaelbn

**Comment by oihane** [comment]
@anajuaristi try this link http://3136482-57-9f1f66.runbot1.odoo-community.org/web?db=3136482-57-9f1f66-base because if you have to activate the BoM you are not using the one @rafaelbn has said, there is not installed mrp_bom_version, the main link is the same but you have to choose between two databases.
@ybofr will you test it also?

**Comment by ybofr** [comment]
Thanks Rafael

**Comment by ybofr** [comment]
Hi,
Please correct the message for the missing routing. It should be "if you haven't assigned a routing yet..."
bill_of_materials_-_odoo

**Comment by ybofr** [comment]
Creation of the BOM. Product unit of measure is fixed on unit and is not taken from the product when the product is created on the fly during BOM editing.
In this case, the metal 1 should come in kg

manufacturing_orders_-_odoo

**Comment by ybofr** [comment]
My comment: test :-1: as the units of measures are not properly taken from the product.

**Comment by pedrobaeza** [comment]
@ybofr, @anajuaristi, the features you are trying are not related to these modules. If you're not going to test what this modules does, please don't review it. The things you're finding should be reported as issues of the already merged modules.

**Comment by JMODaERP** [comment]
Here is the visual process: 

https://docs.google.com/document/d/15d-VkhxKSHbstM7ONXa41KLh5WmixfKu0su0eCTgMQs/edit?ts=565ee128#

**Comment by rafaelbn** [comment]
Resume:
- This PR has 2 x :+1: 
- The 2 x :-1: as commeted by @pedrobaeza are not related as reviewing try to do somethings that this module doesn't do.
- @JMODaERP make a doc to know how to test this module (**thanks a lot for this work**): https://docs.google.com/document/d/15d-VkhxKSHbstM7ONXa41KLh5WmixfKu0su0eCTgMQs/edit?ts=565ee128#
- I will review this module now

If:
- @nhomar is agree with his comments here https://github.com/OCA/manufacture/pull/57#issuecomment-160173847
- @moylop260  is agree with his comments here https://github.com/OCA/manufacture/pull/57#issuecomment-160177905
- @elicoidal cooments reviewed

This module is ready to merge IMHO

**Comment by rafaelbn** [comment]
Tested in runbot :+1: Thanks!

**Comment by anajuaristi** [comment]
@pedrobaeza 
Just an small bug --> confirm the manufacturing order and before starting one operation change the default machine for another one.
Expected behavior: Recalculate cycles and stimated time for Wo
Resulted behavior: cycles and time are not recalculated, remaning the same

**Comment by pedrobaeza** [comment]
OK, let's try to put this in order because there's a lot of noise in this PR:
- @elicoidal, you have made comments in PR https://github.com/OCA/manufacture/pull/60 that belongs here, because as stated in that PR, it includes all this stuff. I have fixed the strings, but I have added a lot more in the README.rst file of _mrp_operations_extension_ module. About `extension` term, I'm afraid in this occasion we cannot use another word nor split the module in parts: it's all attached together as a truly extension for operations. If there are more things needed in operations management, they can be included in the module. Take into account that the boundaries of the module are already very limited.
- @nhomar: I have added tests to increase even more the coverage level. The init hook is safe because it doesn't destroy information. But sorry, no video. I don't like it because they are not "indexable", and I don't have time for make it. I have improved the README file instead.
- @moylop260: I have added images and fixed pylint beta messages.
- @anajuaristi: Try here the error you mention if it still happens.
- @ybofr: The README is improved to see exactly what this module does if you want to test it. There are 2 strategies: remove the other modules of the repo for not interfering, or know what makes each of them for ignoring their behaviours.
- @rafaelbn, @JavierIniesta, @JMODaERP, you can make last test before merging.

File: mrp_operations_extension/README.rst [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/129/8.0
+
+Known issues
+============
+
+* *hr_timesheet* is a big dependency for only catching the employee cost when
+  selectiong operators in the workcenter. A glue module for this operation
```
Comment (elicoidal) [inline mrp_operations_extension/README.rst]: s/selectiong/selecting

**Comment by pedrobaeza** [comment]
@elicoidal, fixed

**Comment by JavierIniesta** [comment]
Runbot is returning a "504 Gateway Time-out" error.

**Comment by pedrobaeza** [comment]
Runbot rebuilding

**Comment by JavierIniesta** [comment]
Tested in local environment :+1:

**Comment by JMODaERP** [comment]
I can jot creat the BoM in this runbot, therefore i can not do the comntrols. Why?. :-1:

**Comment by JMODaERP** [comment]
I can not creat the BoM in this runbot, therefore i can not do the comntrols. Why?.

**Comment by pedrobaeza** [comment]
Because runbot already has _mrp_bom_version_ installed. Uninstall that module or activate the bom.

**Comment by rafaelbn** [comment]
Tested in runbot :+1:

**Comment by pedrobaeza** [comment]
OK, this PR is blocking a lot more, so I'm going to merge with the positive review of 2 reviewers, and with the reasoning behind the thumb down explained. If there's any problem, we can make a PR later.

**Comment by JMODaERP** [comment]
Tested and O.K. again. :+1:

**Pull request merged**